### ISSUE TITLE
msg_id too high

### ISSUE THREAD

**Issue by Ozymandy** [sochix/TLSharp#405]
My program throws exception 
![f9df057d46bb893e7cb3e9f925ef05ee5939079f500a94f932 pimgpsh_fullsize_distr](https://cloud.githubusercontent.com/assets/3194888/23117538/3afd8128-f761-11e6-8e62-a77769ff2745.png)

**Comment by Ozymandy** [comment]
Can it happens because of invalid time on pc. (Local time at the regions and pc time are different)

**Comment by knocte** [comment]
Is that a question or an assertion?

**Comment by Ozymandy** [comment]
It's a question

**Comment by knocte** [comment]
Ok cool. If you end up finding that this is the case, TLSharp should make sure that the clock is in sync before performing any connection.

**Comment by Ozymandy** [comment]
Yes, I found that If your system time more than local time it throws msg_id too high

**Comment by knocte** [comment]
Great, that's a good catch. Can you create a pull request that detects this situation and throws a TimeNotInSyncException() instead of the cryptic msg_id too high?

**Comment by Andrey123** [comment]
This error is quite frequent.

Try increasing the system time by 1 hour, after that the client will not be able to send messages: client.connect() will throw 'msg_id too high'.
Moreover, it will not be able to send messages even if you return the time back.
Moreover, it will not be able to send messages even even if you call reconnect(true) that updates timeOffset, because the session remembers the wrong LastMessageId, which is very large now and all the following messages will have the MessageId even more and will also be rejected. Only reauthorization will help in this case now, otherwise, your account will be banned for 24 hours after the large number of rejected messages (errors).

I learned this question and found the following:

1. The session should not change LastMessageId if the message was rejected.
2. After rejecting, as well as after Salt is changed on the server, the client must request a new TimeOffset and Salt and store them in the session and resend the message with the correct MessageId.
3. After receiving messages from the server, you should check the server message ID, get the server time  (MessageId >> 32) and check if (serverTime > clientTime + 60 || serverTime + 300 < clientTime) then update timeOffset.

Could you please fix this error?

**Comment by knocte** [comment]
> The session should not change LastMessageId if the message was rejected.
> After rejecting, as well as after Salt is changed on the server, the client must request a new TimeOffset and Salt and store them in the session and resend the message with the correct MessageId.
> After receiving messages from the server, you should check the server message ID, get the server time (MessageId >> 32) and check if (serverTime > clientTime + 60 || serverTime + 300 < clientTime) then update timeOffset.

That's a great analysis Andrey, thanks! Pull requests to fix these problems, are welcome.

**Comment by Andrey123** [comment]
Some additional info.

Сlient doesn't need to request a new TimeOffset, it exists in response, for example:

```
private bool HandleMsgsAck(ulong messageId, int sequence, BinaryReader messageReader)
{
	int serverTime = Convert.ToInt32(messageId >> 32);
	int clientTime = Convert.ToInt32(Helpers.LinuxTime.TotalSeconds);

	if (serverTime > clientTime + 60 || serverTime + 300 < clientTime)
	{
		_session.TimeOffset = serverTime - clientTime;
	}

	return false;
}
public static TimeSpan LinuxTime
{
	get { return DateTime.UtcNow - new DateTime(1970, 1, 1); }
}
```
And in case of this error the same code, but without if. But resend is required, and don't change LastMessageId, it's harder....

**Comment by imanf94** [comment]
#586

**Comment by Yashil** [comment]
I have encountered this issue too. Please let me know too if any fix is offered.
